rfc/0003-indexmanagement.md

```diff
@@ -0,0 +1,201 @@
+This choice can be made through a first `boolean` parameter, `ignoreIfExist`. If it is `false`, then a SDK-related exception, `IndexAlreadyExistsException`, should be raised for SDKs that rely on exceptions.
+
+The second specificity can be triggered by another `boolean` parameter, `defer`. If set to true, the N1QL query will use the "with defer" syntax and the index will simply be "pending" (prior to 4.5) or "deferred" (at and after 4.5, see [MB-14679](https://issues.couchbase.com/browse/MB-14679)).
+
+Signatures:
+
+```java
+boolean createPrimaryIndex(boolean ignoreIfExist, boolean defer)
+
+boolean createNamedPrimaryIndex(String customName, boolean ignoreIfExist, boolean defer)
+
+boolean createIndex(String indexName, List<String> fields, boolean ignoreIfExist, boolean defer)
+```
+
+For languages that permit it, the list of fields in `createIndex` could also accept `Expression` objects (or any relevant equivalent). Rather than interpreting the entry as a field name to be escaped, the entry could then be used "as is", allowing for more complex constructions (like an index defined as "`DISTINCT ARRAY v FOR v IN schedule END`" for instance).
+
+## Dropping Indexes
+For index removal, the inverse of `ignoreIfExist` must be put in place: N1QL considers it an error to attempt dropping an index that doesn't exist. So the `ignoreIfNotExist` will drive the behavior of the API in such a case: either `false` and an `IndexDoesNotExistException` should be thrown (or equivalent for SDKs that don't rely on exceptions), or `true` and the invocation will do nothing.
+
+Signatures:
+
+```java
+boolean dropPrimaryIndex(boolean ignoreIfNotExist)
+
+boolean dropNamedPrimaryIndex(String customName, boolean ignoreIfNotExist)
+
+boolean dropIndex(String name, boolean ignoreIfNotExist)
+```
+
+> **Note**: Unlike index creation, deferring the drop of an index is not possible
+
+## Building deferred indexes
+Signature:
+
+```java
+List<String> buildDeferredIndexes()
+```
+
+Remarks:
+The returned list is the list of indexes that were in **`pending`** or **`deferred`** state at the time of the call.
+
+The method will first list indexes, only keeping the `pending`/`deferred` ones. It then issues the appropriate N1QL statement from the list of pending indexes (here `INDEX_NAME1` and `INDEX_NAME3`):
+
+```sql
+BUILD INDEX ON `BUCKET_NAME_INJECTED_HERE`(`INDEX_NAME1`, `INDEX_NAME3`)
+```
+
+# Watching Indexes
+>**Note**: this part of the API could be considered optional.
+
+Since the process of building a deferred index is asynchronous, the server immediately returns control to the client. It would be useful to have an API to poll the list of indexes until one (or several) indexes are seen as "online". In the blocking API, this would allow to make sure that an index is ready.
+
+Signature:
+
+```java
+List<IndexInfo> watchIndexes(List<String> watchList, boolean watchPrimary, long watchTimeout, TimeUnit watchTimeUnit)
```

Comment:
I am relatively sure, but it's certainly something we can bring up with the query team.  In regards to your first statement, even if they didn't defer-build the primary index, it would still need to be built before secondary indexes could function, so there is no downside to ensuring it is built (if its not deferred, its going to be in the process of building by default anyways).